### PR TITLE
Add generic Dispatcher service to zocalo

### PR DESCRIPTION
**Pull request by TimGuiteDiamond**
Can form basis for site specific Dispatchers which take care of their own logging, recipe base paths etc

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -47,7 +47,8 @@
         ],
         "libtbx.precommit": ["zocalo = zocalo"],
         "workflows.services": [
-            "Schlockmeister = zocalo.service.schlockmeister:Schlockmeister"
+            "Schlockmeister = zocalo.service.schlockmeister:Schlockmeister",
+            "Dispatcher = zocalo.service.dispatcher:Dispatcher",
```
Comment (Anthchirp) [inline setup.py]: please swap, so that they're alphabetically ordered

File: zocalo/service/dispatcher.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    _service_name = "Dispatcher"
+
+    # Logger name
+    _logger_name = "services.dispatcher"
+
+    # Define a base path where your recipes are located (accessed with zocalo.go -r $recipename)
+    _recipe_basepath = None
+
+    # Define a logbook if you would like to do logging
+    _logbook = None
```
Comment (Anthchirp) [inline zocalo/service/dispatcher.py]: I think everything lookbook related should be left out. Replace with _before-filter_ and (_after-filter_ or _before-dispatch_) hook.

File: zocalo/service/dispatcher.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+    def filter_messages(self, message, parameters):
+        """
+        Filter out the messages and parameters and return them as a tuple.
+        This can be implemented in child classes with different filters depending
+        on the site and specific implementation
+        """
+        return (message, parameters)
```
Comment (Anthchirp) [inline zocalo/service/dispatcher.py]: maybe `_message_filters = []` would be better? That allows using multiple filters.

code further down then becomes
```
for f in self._message_filters:
  message, parameters = f(message, parameters)
```

File: zocalo/service/dispatcher.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+                "Dispatcher rejected malformed message: parameters not given as dictionary"
+            )
+            self._transport.nack(header)
+            return
+
+        # Unless 'guid' is already defined then generate a unique recipe IDs for
+        # this request, which is attached to all downstream log records and can
+        # be used to determine unique file paths.
+        recipe_id = parameters.get("guid") or str(uuid.uuid4())
+        parameters["guid"] = recipe_id
```
Comment (Anthchirp) [inline zocalo/service/dispatcher.py]: suggest a search/replace `guid` -> `uuid`. Best opportunity we have to fix the naming.

File: zocalo/service/dispatcher.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+
+        # If we are fully logging requests then make a copy of the original message
+        if self._logbook:
+            original_message = copy.deepcopy(message)
```
Comment (Anthchirp) [inline zocalo/service/dispatcher.py]: replace with a _before-filter_-hook

File: zocalo/service/dispatcher.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+                recipe=full_recipe, transport=self._transport
+            )
+            rw.environment = {
+                "ID": recipe_id
+            }  # FIXME: This should go into the constructor, but workflows can't do that yet
+            rw.start(transaction=txn)
+
+            # Write information to logbook if applicable
+            if self._logbook:
+                self.record_to_logbook(recipe_id, header, original_message, message, rw)
```
Comment (Anthchirp) [inline zocalo/service/dispatcher.py]: replace with a _after-filter_ or _before-dispatch_-hook

File: zocalo/service/dispatcher.py [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+            # Conditionally acknowledge receipt of the message
+            txn = self._transport.transaction_begin()
+            self._transport.ack(header, transaction=txn)
+
+            rw = workflows.recipe.RecipeWrapper(
+                recipe=full_recipe, transport=self._transport
+            )
+            rw.environment = {
+                "ID": recipe_id
+            }  # FIXME: This should go into the constructor, but workflows can't do that yet
```
Comment (Anthchirp) [inline zocalo/service/dispatcher.py]: I've just made that change, so can replace the 6 lines with
```python
rw = workflows.recipe.RecipeWrapper(
                environment={"ID": recipe_id},
                recipe=full_recipe,
                transport=self._transport,
            )
```

File: zocalo/cli/go.py [hunk 8]
```diff
@@ -37,7 +37,8 @@ def run():
         metavar="RCP",
         action="append",
         default=[],
-        help="Name of a recipe to run. Can be used multiple times. Recipe names correspond to filenames (excluding .json) in /dls_sw/apps/zocalo/live/recipes",
+        help="Name of a recipe to run. Can be used multiple times. Recipe names correspond to filenames (excluding .json)."
+        " This directory is defined in the _recipe_basepath attribute of Dispatcher services.",
```
Comment (Anthchirp) [inline zocalo/cli/go.py]: assert this information is not helpful for anyone running `zocalo.go --help`

File: zocalo/service/dispatcher.py (hunk 2 above)
Comment (TimGuiteDiamond) [reply zocalo/service/dispatcher.py]: Should these hooks take arguments?

File: zocalo/service/dispatcher.py (hunk 3 above)
Comment (TimGuiteDiamond) [reply zocalo/service/dispatcher.py]: Yep, done

File: zocalo/service/dispatcher.py (hunk 4 above)
Comment (TimGuiteDiamond) [reply zocalo/service/dispatcher.py]: Done, I think? Will this cause compatibility errors?

File: zocalo/service/dispatcher.py (hunk 7 above)
Comment (TimGuiteDiamond) [reply zocalo/service/dispatcher.py]: Done

File: zocalo/cli/go.py (hunk 8 above)
Comment (TimGuiteDiamond) [reply zocalo/cli/go.py]: Okay, suggestions?

Comment (TimGuiteDiamond) [reply zocalo/cli/go.py]: Presumably go could be run without having this defined if you only used -f ? Currently the message is only helpful if you use DLSDispatcher at Diamond

File: zocalo/service/dispatcher.py (hunk 4 above)
Comment (Anthchirp) [reply zocalo/service/dispatcher.py]: only in dlstbx, and these need to be addressed when we change DLSDispatcher to inherit from Dispatcher

**Comment by Anthchirp** [comment]
I just realized that the file lookup mechanism is really just another filter, so it's probably worth thinking a bit more about the structure this should have.

**Comment by TimGuiteDiamond** [comment]
> I just realized that the file lookup mechanism is really just another filter, so it's probably worth thinking a bit more about the structure this should have.

The structure of the whole dispatcher or just of the filtering?

**Comment by TimGuiteDiamond** [comment]
It does seem that go and dispatcher have quite a lot of dependencies on each other, could they be turned into one thing?
I.e does Dispatcher have to be a service? Could the filtering be done during the go command? This would mean having to define your own go command (instead of your own Dispatcher subclass) but could provide finer control over what filtering occurs? Perhaps some go commands should be filtered with different filters than others? This is not possible with the current set up because if there were multiple different Dispatchers there is no control over which one will process the request.

**Comment by Anthchirp** [comment]
I was going to say "just merge it, can sort out any issues later", but the custom_recipe -> raw_recipes change breaks compatibility between zocalo.go and DLSDispatcher, which I believe I missed in my initial review.
I'm not entirely sure what the use case for submitting _n_ independent recipes at the same time is? Wouldn't one be better off submitting _n_ messages?

**Comment by stufisher** [comment]
Superseded by https://github.com/DiamondLightSource/python-zocalo/pull/118

**Pull request closed without merging**